linera-execution/src/wasm/wasmer.rs

```diff
@@ -273,27 +267,65 @@ impl From<wasmer::RuntimeError> for ExecutionError {
 }
 
 /// Serialized bytes of a compiled contract bytecode.
-///
-/// Each [`Module`] needs to be compiled with a separate [`Engine`] instance, otherwise Wasmer
-/// [panics](https://docs.rs/wasmer-middlewares/3.3.0/wasmer_middlewares/metering/struct.Metering.html#panic)
-/// because fuel metering is configured and used across different modules.
 pub struct CachedContractModule {
     compiled_bytecode: Bytes,
 }
 
+pub fn add_metering(bytecode: Bytecode) -> anyhow::Result<Bytecode> {
+    struct WasmtimeRules;
+
+    impl gas_metering::Rules for WasmtimeRules {
+        /// Calculates the fuel cost of a WebAssembly [`Operator`].
+        ///
+        /// The rules try to follow the hardcoded [rules in the Wasmtime runtime
+        /// engine](https://docs.rs/wasmtime/5.0.0/wasmtime/struct.Store.html#method.add_fuel).
+        fn instruction_cost(
+            &self,
+            instruction: &parity_wasm::elements::Instruction,
+        ) -> Option<u32> {
+            use parity_wasm::elements::Instruction::*;
+
+            Some(match instruction {
+                Nop | Drop | Block(_) | Loop(_) | Unreachable | Else | End => 0,
+                _ => 1,
+            })
+        }
+
+        fn memory_grow_cost(&self) -> gas_metering::MemoryGrowCost {
+            gas_metering::MemoryGrowCost::Free
+        }
+
+        fn call_per_local_cost(&self) -> u32 {
+            0
+        }
+    }
+
+    let instrumented_module = gas_metering::inject(
+        parity_wasm::deserialize_buffer(&bytecode.bytes)?,
+        gas_metering::mutable_global::Injector::new(GAS_GLOBAL_NAME),
+        &WasmtimeRules,
+    )
+    .map_err(|_| anyhow::anyhow!("failed to instrument module"))?;
+
+    Ok(Bytecode {
+        bytes: instrumented_module.into_bytes()?,
+    })
+}
+
 impl CachedContractModule {
     /// Creates a new [`CachedContractModule`] by compiling a `contract_bytecode`.
     pub fn new(contract_bytecode: Bytecode) -> Result<Self, anyhow::Error> {
-        let module = Module::new(&Self::create_compilation_engine(), contract_bytecode)?;
+        let module = Module::new(
+            &Self::create_compilation_engine(),
```

Comment:
Minor: we might be able to share the engine now. Could be worth a try, because it could improve performance a bit.

Reply:
When I've got it passing tests I'll give it a go!

Reply:
On second thoughts I'd quite like to do this in a separate PR :thinking: